File: jsset/GameOfSet.js
/**
 * Main code 
 */
let game = new GameOfSet();


/**
 * Card class
 */

/**
 * Constructor for card object
 */
function Card(number, color, shading, shape, imageName){
    this.number = number;
    this.color = color;
    this.shading = shading;
    
    this.shape = shape;
    this.id = number + color + shading + shape;
    this.imageName = imageName;

    this.tag = cardTag(number, color, shading, shape, imageName);
    this.tag.select = select;
}

function cardTag(number, color, shading, shape, imageName){
    var imageTag = document.createElement("img");
    imageTag.setAttribute("src", imageName);
    var elementTag = document.createElement("td");
    elementTag.setAttribute("id", number + color + shading + shape);
    elementTag.setAttribute("class", "not-selected");
    elementTag.setAttribute("onclick", "this.select()");
    elementTag.appendChild(imageTag);
    
    return elementTag;
}



function select(){

    if (this.getAttribute("class") == "selected"){
        this.setAttribute("class", "not-selected");
    } else {
        this.setAttribute("class", "selected");
    }


}

/**
 * Method for creating the deck.
 */

function shuffleDeck(deck){
    for (var i = deck.length - 1; i > 0; i--) {
        var j = Math.floor(Math.random() * (i + 1));
        var temp = deck[i];
        deck[i] = deck[j];
        deck[j] = temp;
    }
}

function createDeck(){
    let deck = new Array(81);
    var index = 0;

    var number = ["one", "two", "three"];
    var color = ["green", "blue", "pink"];
    var shading = ["open", "shaded", "solid"];
    var shape = ["oval", "diamond", "square"];

    for (let w = 0; w < 3; w++){
        for (let x = 0; x < 3; x++){
            for (let y = 0; y < 3; y++){
                for (let z = 0; z < 3; z++){
                    var imageName = "card_images/" + number[w] + "_" + color[x] + "_" + shading[y] + "_" + shape[z] + ".png"; 
                    deck[index] = new Card(number[w], color[x], shading[y], shape[z], imageName);
                    index++;
                }
            }
        }
    }


    shuffleDeck(deck);

    return deck;
}

/**
 *  Creates initial dealt card set
 */

function dealCards(deck){
    let dealt = new Array();
    for (let i = 0; i < 12; i++){
        dealt.push(deck.pop());
    }


    return dealt;
}

/**
 * Updates cards displayed on webpage
 */

function display(dealt){
    var row1 = document.getElementById("row1");
    var row2 = document.getElementById("row2");
    var row3 = document.getElementById("row3");

    for (let i = 0; i < dealt.length; i++){

        if (i % 3 == 0){
            row1.appendChild(dealt[i].tag);
        } else if (i % 3 == 1) {
            row2.appendChild(dealt[i].tag);
        } else {
            row3.appendChild(dealt[i].tag);
        }
    }
}

/**
 * GameOfSet class
 */

function GameOfSet(){
    this.deck = createDeck();
    this.dealt = dealCards(this.deck);
    this.score1 = 0;
    this.score2 = 0;
    display(this.dealt);

}

/**
 * Game engine functions
 */

/**
 * Checks if game is over
 */

function gameOver(){
    return (!hasSet(game.dealt) && game.deck.length == 0);
}

/**
 * Declares winner
 */

function endGame(){
    if (game.score1 > game.score2){
        window.alert("GAME OVER: Player 1 Wins!!!");
    } else if (game.score1 < game.score2){
        window.alert("GAME OVER: Player 2 Wins!!!");
    } else {
        window.alert("GAME OVER: It's a Tie!!!");
    }
}

/**
 * Checks if a given set is a valid set
 */

 function checkSet(cards){
	// length 3 arrays for each attribute of the three Cards
	    let number = [cards[0].number, cards[1].number, cards[2].number];
    	let color = [cards[0].color, cards[1].color, cards[2].color];
    	let shading = [cards[0].shading, cards[1].shading, cards[2].shading];
    	let shape = [cards[0].shape, cards[1].shape, cards[2].shape];

    	//change to array of set to remove duplicates
    	let uniqueNumber = new Set(number);
    	let uniqueColor = new Set(color);
    	let uniqueShading = new Set(shading);
    	let uniqueShape = new Set(shape);

    	return !(uniqueNumber.size == 2 || uniqueColor.size == 2 || uniqueShading.size == 2 || uniqueShape.size == 2);
}

/**
 * Checks if there is a valid set in the dealt cards
 */

 function hasSet(dealt){
	//put all of the possible combinations of 3 from dealt into an array
    var possibleSets = new Array();
    var actualSets = new Array();

	for (let i = 0; i < dealt.length; i++){
    		for (let j = i + 1; j < dealt.length; j++){
    			for (let x = j + 1; x < dealt.length; x++){
        			possibleSets.push([dealt[i], dealt[j], dealt[x]]);
        		}
        	}
        }
        for (let i = 0; i < possibleSets.length; i++){ 
        	if (checkSet(possibleSets[i])){
        		actualSets.push(possibleSets[i]);
        	}
        }
        return actualSets.length > 0
}

/**
 * Number of selected cards
 */

function numSelected(){
    var count = 0;
    for (let i = 0; i < game.dealt.length; i++){
        if (game.dealt[i].tag.getAttribute("class") == "selected"){
            count++;
        }
    }

    return count;
}

/**
 * Array of cards selected
 */

function getSelectedCards(){

    let selectedCards = new Array();
    for (let i = 0; i < game.dealt.length; i++){
        if (game.dealt[i].tag.getAttribute("class") == "selected"){
            selectedCards.push(game.dealt[i]);
        }
    }

    return selectedCards;
}

/**
 * Removes a set of cards without replacing
 */

function removeCards(cards){
    for (let i = 0; i < 3; i++){
        cards[i].tag.parentNode.removeChild(cards[i].tag);
    }

    game.dealt = game.dealt.filter(card => card.tag.getAttribute("class") != "selected");
}

/**
 * Replaces a set of cards on screen
 */

function replaceCards(cards){
    game.dealt = game.dealt.filter(card => card.tag.getAttribute("class") != "selected");
    
    for (let i = 0; i < 3; i++){
        let newCard = game.deck.pop();
        cards[i].tag.parentNode.appendChild(newCard.tag);
        game.dealt.push(newCard);
        cards[i].tag.parentNode.removeChild(cards[i].tag);
    }
}

/**
 * Adds cards if allowed when user pushes ADD button
 */

function addCards(){
    if (gameOver()){
        window.alert("Game has ended! Refresh the page to play again!");
    } else if (!hasSet(game.dealt)){
    var row1 = document.getElementById("row1");
    var row2 = document.getElementById("row2");
    var row3 = document.getElementById("row3");

    var new1 = game.deck.pop();
    var new2 = game.deck.pop();
    var new3 = game.deck.pop();

    game.dealt.push(new1);
    game.dealt.push(new2);
    game.dealt.push(new3);

    row1.appendChild(new1.tag);
    row2.appendChild(new2.tag);
    row3.appendChild(new3.tag);
    } else if (game.deck.length == 0) {
        window.alert("Cannot add cards. There are no more cards in the deck!");
    } else {
        window.alert("Cannot add cards. There is a set in the dealt cards!");
    }
}

/**
 * Gives user a hint on request
 */

function hint(){
    var dealt = game.dealt;
    if (gameOver()){
        window.alert("Game has ended! Refresh the page to play again!");
    } else if (hasSet(dealt)){
        var possibleSets = new Array();
    var actualSets = new Array();

	for (let i = 0; i < dealt.length; i++){
    		for (let j = i + 1; j < dealt.length; j++){
    			for (let x = j + 1; x < dealt.length; x++){
        			possibleSets.push([dealt[i], dealt[j], dealt[x]]);
        		}
        	}
        }
        for (let i = 0; i < possibleSets.length; i++){ 
        	if (checkSet(possibleSets[i])){
        		actualSets.push(possibleSets[i]);
        	}
        }
        actualSets[0][0].tag.setAttribute("class", "selected");
    } else {
        window.alert("No valid sets present!");
    }
}

/**
 * Controller and event functions
 */

/**
 * User calls "set"
 */

function setButton(){
    if (gameOver()){
        window.alert("Game has ended! Refresh the page to play again!");
    } else if (numSelected() == 3){
        let selectedCards = getSelectedCards();
        if (checkSet(selectedCards)) {
            game.score1++;
            if (game.dealt.length > 12 || game.deck.length == 0){
                removeCards(selectedCards);
            } else {
               replaceCards(selectedCards);
            }
            if (gameOver()){
                endGame();
            }
        } else {
            if (game.score1 > 0){
                game.score1--;
            }
            window.alert("Not a valid set!");
        }
        updateScore();
    } else {
        window.alert("Must have three cards selected to check for a set!");
    }
}

function setButton2(){
    if (gameOver()){
        window.alert("Game has ended! Refresh the page to play again!");
    } else if (numSelected() == 3){
        let selectedCards = getSelectedCards();
        if (checkSet(selectedCards)) {
            game.score2++; 
            if (game.dealt.length > 12 || game.deck.length == 0){
                removeCards(selectedCards);
            } else {
               replaceCards(selectedCards);
            }
            if (gameOver()){
                endGame();
            }
        } else {
            if (game.score2 > 0){
                game.score2--;
            }
            
            window.alert("Not a valid set!");
        }
        updateScore();
    } else {
        window.alert("Must have three cards selected to check for a set!");
    }
}

/**
 * Update score on screen
 */

function updateScore(){
    var p1 = document.getElementById("player1Score");
    var p2 = document.getElementById("player2Score");

    p1.firstChild.nodeValue = "Player1: " + game.score1;
    p2.firstChild.nodeValue = "Player2: " + game.score2;
}

File: rubyset/GameOfSet.rb
# GameOfSet.rb: represents an instance of the Game of Set for GUI

require 'ruby2d'
require_relative 'GUICard.rb'
require_relative 'GUIDeck.rb'

class GameOfSet
  
  # intialize 81 card deck using GUIDeck class method
  @@deck = createDeck

  # array that will hold the cards currently dealt/visible
  @@dealt = Array.new

  # setup game window 
  def initialize

    # start with 12 cards. pop them off the deck and reposition to fill three
    # rows. add image to screen to make visible
    for i in (0...12)
      @@dealt[i] = @@deck.pop
      @@dealt[i].reposition((i - 4 * (i/4)) * 96, (i/4) * 135)
      @@dealt[i].image.add
    end

    # display instructions below dealt cards
    @instructions = Array.new(5)
    @instructions[0] = Text.new(
      "Left click on cards to select/deselect.",
      x: 15, y: 420,
      size: 12,
      color: 'black'
    )
    @instructions[1] = Text.new(
      "Press 'a' key to add cards.",
      x: 15, y: 440,
      size: 12,
      color: 'black'
    )
    @instructions[2] = Text.new(
      "Player1: press 's' key to check set.",
      x: 15, y: 460,
      size: 12,
      color: 'black'
    )
    @instructions[3] = Text.new(
      "Player2: press 'k' key to check set.",
      x: 15, y: 480,
      size: 12,
      color: 'black'
    )
    @instructions[4] = Text.new(
      "Press 'h' key for hint. Must have no cards selected.",
      x: 15, y: 500,
        size: 12,
        color: 'black'
    )

    # create the helper. displays tips on bottom of window when user exhibits
    # unexpected behavior (i.e. attempting to check a set with only two cards
    # selected)
    @helper = Text.new(
      "Help:",
      x: 15, y: 525,
      size: 12,
      color: 'red',
      style: 'bold'
    )

    # player scores start at zero
    @p1score = 0
    @p2score = 0

    # display for each player's score
    @scoreboard = Array.new(2)
    @scoreboard[0] = Text.new(
      "Player1: 0",
      x: 500, y: 430,
      size: 15,
      color: 'black',
      style: 'bold'
    )
    @scoreboard[1] = Text.new(
      "Player2: 0",
      x: 500, y: 460,
      size: 15,
      color: 'black',
      style: 'bold'
    )

  end

  # selects cards based on given mouse input x and y
  def selector(x, y)
    for i in (0...@@dealt.length)
      if @@dealt[i].image.contains? x , y
        @@dealt[i].select
      end
    end

  end

  # determines if a given set of Cards of size three is an acceptable set 
  # based on the rules of the game
  def set?(cards)

    # length 3 arrays for each attribute of the three Cards
    number = [cards[0].number, cards[1].number, cards[2].number]
    color = [cards[0].color, cards[1].color, cards[2].color]
    shading = [cards[0].shading, cards[1].shading, cards[2].shading]
    shape = [cards[0].shape, cards[1].shape, cards[2].shape]

    # remove duplicates
    number.uniq!
    color.uniq!
    shading.uniq!
    shape.uniq!

    # attributes must be all same or all different. if one array has a length
    # of two, then there are two same and one different, which is not allowed
    return !(number.length == 2 || color.length == 2 || shading.length == 2 || shape.length == 2)
  end

  # checks if current dealt cards contain a valid set
  def hasSet?

    # store all possible combinations of three of the Cards dealt
    possibleSets = @@dealt.combination(3).to_a
    # store all valid sets from previous array in new array
    actualSets = possibleSets.select {|cards| set?(cards)}
    # length will be greater than zero if a valid set combination is found
    return actualSets.length > 0
  end

  # selects one card from one of the valid sets in the dealt cards
  def hint
    possibleSets = @@dealt.combination(3).to_a
    actualSets = possibleSets.select {|cards| set?(cards)}

    actualSets[0][0].select
  end

  # determines the number of cards on the screen currently selected
  def selectedCards
    return @@dealt.select {|card| card.selected}
  end

  # adds three more cards if no valid set is present in dealt cards
  def addCards
    @@dealt.push @@deck.pop
    @@dealt.push @@deck.pop
    @@dealt.push @@deck.pop
    
    updateDisplay
  end

  # replaces cards after a set is found by user and removed by the program
  def replaceCards(cards)
    for i in (0..2)
      @@dealt[@@dealt.index(cards[i])] = @@deck.pop
      cards[i].clear
    end
    updateDisplay
  end

  # removes cards from dealt array; for when there are more than 12 cards on
  # the screen or when the deck is empty
  def removeCards(cards) 
    for i in (0...cards.length)
      cards[i].clear
      @@dealt.delete(cards[i])
    end
    updateDisplay
  end

  # repositions the cards currently dealt in three rows
  def updateDisplay
    length = @@dealt.length
    columns = length/3
    for i in (0...length)
      @@dealt[i].reposition((i - columns*(i/columns)) * 96, (i/columns) * 135)
      @@dealt[i].image.add
    end
  end

  # changes what is displayed by the helper text
  def updateHelper(string)
    @helper.remove
    @helper = Text.new(
      'Help: ' + string,
      x: 15, y: 525,
      size: 12,
      color: 'red',
      style: 'bold'
    )
  end

  # updates player scores; for when set is found or wrong attempt tried
  def updateScore(player, correct)
    case player
    when 1
      if correct
        @p1score += 1
      elsif !correct && @p1score != 0
        @p1score -= 1
      end
    when 2
      if correct
        @p2score += 1
      elsif !correct && @p2score != 0
        @p2score -= 1
      end
    end

    updateScoreboard
  end

  # updates the displayed scores on the screen
  def updateScoreboard
    @scoreboard[0].remove
    @scoreboard[0] = Text.new(
      "Player1: " + @p1score.to_s,
      x: 500, y: 430,
      size: 15,
      color: 'black',
      style: 'bold'
    )

    @scoreboard[1].remove
    @scoreboard[1] = Text.new(
      "Player2: " + @p2score.to_s,
      x: 500, y: 460,
      size: 15,
      color: 'black',
      style: 'bold'
    )
  end

  # when the deck is empty and there are no more sets present in the dealt
  # cards, the game is over
  def endGame?
    return @@deck.length == 0 && !hasSet?
  end

  # clear remaining cards, determine winnner and display final message to user
  def endGame
    winner = ""
    for card in @@dealt
      card.clear
    end

    @@dealt.clear

    if @p1score > @p2score
      winner = "Player1 wins!"
    elsif @p2score > @p1score
      winner = "Player2 wins!"
    else
      winner = "It's a tie!"
    end

    finalMessage = Text.new(
      winner,
      x: 320, y: 250,
      size: 20,
      color: 'red',
      style: 'bold'
    )
  end

  # number of cards remaining in deck
  def deckLength
    @@deck.length
  end

  # number of cards currently on the screen
  def dealtLength
    @@dealt.length
  end
end

